### PR TITLE
Add Foreign instances for HugeNum.

### PR DESCRIPTION
**Pull request by chexxor**
I'd like to define an `IsForeign` instance for one my data type which uses HugeNum. How does this look? Should I remove the `derive instance genericDigit :: Generic Digit` addition?

### CODE REVIEW

File: src/Data/HugeNum.purs [hunk 1]
```diff
@@ -73,6 +74,14 @@ timesSign Plus Plus = Plus
 timesSign Minus Minus = Plus
 timesSign _ _ = Minus
 
+instance isForeignHugeNum :: IsForeign HugeNum where
+  read f = do
+    n <- readNumber f
+    pure (fromNumber n)
+
+instance asForeignHugeNum :: AsForeign HugeNum where
+  write v = write (toNumber v)
```
Comment (garyb) [inline src/Data/HugeNum.purs]: `HugeNum`s are not foreign (in that they're not a type that "comes from JavaScript"), so I don't think we should provide these instances. Also seems like these instances would only work for `HugeNum`s in the range of `Number`, which seems a bit redundant.

We could provide `EncodeJson` / `DecodeJson` instances for argonaut if the idea is to have these for serialisation.

**Comment by garyb** [comment]
Making `Digit` generic is fine by me 👍

File: src/Data/HugeNum.purs (hunk 1 above)
Comment (chexxor) [inline src/Data/HugeNum.purs]: You make good points - the IsForeign instance would only work for a subset of values of HugeNum. I don't see any laws/expections defining IsForeign. Is zero information loss a presumed law? Only JS-native types can be IsForeign? I only see this statement about its presumed law: "_safely_ coerce" - I wonder if this should be updated.

Comment (garyb) [inline src/Data/HugeNum.purs]: Yeah, I think generally we need to improve the documentation of `Foreign`, as it's not clear that it's intended for interop with JS, not arbitrary serialization of JS and JSON / JSON parsing.

I think making foreign have laws in the usual sense would be a bit tricky, but yeah, definite room for improvement.

Going way off topic here, but I guess the best way I can describe it is this: when using `Foreign` there should be no change at all in the representation of the value (null/undefined being maybe-ified aside), it's there to verify a value fits an expected representation.

Comment (chexxor) [inline src/Data/HugeNum.purs]: Roger that. Sounds good. It's good to have this discussion here as documentation, as I'm sure others will be curious about this. I'll close - thanks Gary!

**Pull request closed without merging**